collector/receiver/telemetryapireceiver/receiver.go

```diff
@@ -196,9 +216,106 @@ func (r *telemetryAPIReceiver) createTraces(slice []telemetryapi.Event) (ptrace.
 	return ptrace.Traces{}, errors.New("no traces created")
 }
 
+func (r *telemetryAPIReceiver) createMetrics(slice []telemetryapi.Event) (pmetric.Metrics, error) {
+	metrics := pmetric.NewMetrics()
+	resourceMetric := metrics.ResourceMetrics().AppendEmpty()
+	r.resource.CopyTo(resourceMetric.Resource())
+	scopeMetric := resourceMetric.ScopeMetrics().AppendEmpty()
+	scopeMetric.Scope().SetName(scopeName)
+	for _, el := range slice {
+		r.logger.Debug(fmt.Sprintf("Event: %s", el.Type), zap.Any("event", el))
+		if r.metricsStartTime.IsZero() {
+			if t, err := parseTimestamp(el.Time); err == nil {
+				r.metricsStartTime = t
+			} else {
+				return pmetric.Metrics{}, err
+			}
+
+		}
+		switch el.Type {
+		case string(telemetryapi.PlatformInitReport):
+			jsonStr, err := json.Marshal(el.Record)
+			if err != nil {
+				return pmetric.Metrics{}, err
+			}
+			var report platformInitReport
+			if err := json.Unmarshal(jsonStr, &report); err != nil {
+				return pmetric.Metrics{}, err
+			} else {
+				if report.Phase == initPhaseInit {
+					r.coldStartCounter++
+					metrics := scopeMetric.Metrics().AppendEmpty()
+					metrics.Metadata().PutStr("type", el.Type)
+					metrics.SetName(semconv.AttributeFaaSColdstart)
+					sum := metrics.SetEmptySum()
+					sum.SetIsMonotonic(true)
+					sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+					dp := sum.DataPoints().AppendEmpty()
+					dp.SetIntValue(r.coldStartCounter)
+					dp.SetStartTimestamp(pcommon.NewTimestampFromTime(r.metricsStartTime))
+					dp.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+					dp.Attributes().PutStr(semconv.AttributeFaaSTrigger, semconv.AttributeFaaSTriggerOther)
+				}
+			}
+		case string(telemetryapi.PlatformReport):
+			r.invocationsCounter++
+			metrics := scopeMetric.Metrics().AppendEmpty()
+			metrics.Metadata().PutStr("type", el.Type)
+			metrics.SetName("faas.invocations")
+			sum := metrics.SetEmptySum()
+			sum.SetIsMonotonic(true)
+			sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+			dp := sum.DataPoints().AppendEmpty()
+			dp.SetIntValue(r.invocationsCounter)
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(r.metricsStartTime))
+			dp.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+			dp.Attributes().PutStr(semconv.AttributeFaaSTrigger, semconv.AttributeFaaSTriggerOther)
```

Comment:
There are some parts of each `case` and within the `if-else` below that repeat a bit, though not completely. Would it make sense to factor our some helper functions, or would context be lost and/or too un-readable>

Reply:
Sure. Done